### PR TITLE
Initial Scala 3 support for jsoniter-scala-macros

### PR DESCRIPTION
**Pull request by plokhotnyuk**
It is the amazing work of @rssh that adds an initial Scala support for jsoniter-scala-macros module (Macros API of jsoniter-scala).

Please see an initial discussion here: https://github.com/plokhotnyuk/jsoniter-scala/issues/236#issuecomment-997179484

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -157,12 +159,19 @@ lazy val `jsoniter-scala-macros` = crossProject(JVMPlatform, JSPlatform)
   .settings(commonSettings)
   .settings(publishSettings)
   .settings(
-    crossScalaVersions := Seq("2.13.7", "2.12.15", "2.11.12"),
-    libraryDependencies ++= Seq(
-      "org.scala-lang" % "scala-compiler" % scalaVersion.value,
-      "org.scala-lang" % "scala-reflect" % scalaVersion.value,
-      "com.beachape" %%% "enumeratum" % "1.6.1" % Test,
-      "org.scalatest" %%% "scalatest" % "3.2.10" % Test
+    crossScalaVersions := Seq("3.1.0","2.13.7", "2.12.15", "2.11.12"),
```
Comment (domdorn) [inline build.sbt:162]: hmm.. would be surprised if the scala macros would build on  < 3.0 versions?

Comment (rssh) [reply build.sbt:162]: currently the old code in 2.13.7 directory.  Probably it should be duplicated in 2.12.5 and 2.11.2. and then we can simplify 2.x versions.

Comment (plokhotnyuk) [reply build.sbt:162]: Probably, the old version of `JsonCodecMaker.scala` need to be moved into `scala-2` (not `scala-2.13` as proposed currently) to be accessible for 2.12 and 2.11 versions of cross-build too.

Comment (plokhotnyuk) [reply build.sbt:162]: @rssh That could be managed out by renaming `scala-2.13` from `jsoniter-scala-macros/shared/src/main` to `scala-2` and adding it to the list of unmanaged sources for the jsoniter-scala-macros module:
```
    Compile / unmanagedSourceDirectories ++=
      CrossVersion.partialVersion(scalaVersion.value) match {
        case Some((2, _)) => CrossType.Full.sharedSrcDir(baseDirectory.value, "main").toList.map(f => file(f.getPath + "-2"))
        case _  => Nil
      },
```

**Comment by rssh** [comment]
Huge milestone --- tests are running.
(now 37 failed,  159 successes, failing mostly because of different error messages;  scala3 enum handling is not written yet, and so on.  Willl return to this  later).

File: build.sbt (hunk 1 above)
Comment (rssh) [reply build.sbt:172]: applied

**Comment by rssh** [comment]
I think it's more or less ready for integration.  At least, all existing tests run successfully.

Some things, which still left (but can be done later) is
- add tests for java enums using in 
- document the new behavior of 'printCodec'. (since Scala3 has no -Xmacro-settings, I introduced JsonCodeMakerSettings.PrintCodec flag,  existence of implicit JsonCodeMakerSettings.PrintCodec in current scope is checked instead settings [btw, I submitted path for -Xmacro-settings to scala3 compiler (https://github.com/lampepfl/dotty/pull/14234), but the acceptance depends on epfl which is out of my control, so better not to rely on it])
- thinking about new flags for using scala3 enums  ordering as discriminator (and write more test-cases)
- more tests and operations for mini-interpreter of constant values.
- more code cleanup

What good, that those activities can be done incrementally when needed.

**Pull request merged**